File: app/src/main/java/github/daisukikaffuchino/rebootnya/fragment/HomeFragment.java
package github.daisukikaffuchino.rebootnya.fragment;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.Button;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;

import github.daisukikaffuchino.rebootnya.MainActivity;
import github.daisukikaffuchino.rebootnya.R;
import github.daisukikaffuchino.rebootnya.SettingsActivity;
import github.daisukikaffuchino.rebootnya.shizuku.NyaShellManager;
import github.daisukikaffuchino.rebootnya.utils.ListItemEnum;
import github.daisukikaffuchino.rebootnya.utils.NyaSettings;
import github.daisukikaffuchino.rebootnya.utils.NyaUtilKt;
import github.daisukikaffuchino.rebootnya.utils.RootUtil;
import github.daisukikaffuchino.rebootnya.utils.ShizukuUtil;

public class HomeFragment extends DialogFragment {
    private Context context;
    private int checkedItem = 0;
    private RootUtil rootUtil;
    private ShizukuUtil shizukuUtil;

    @SuppressLint("InflateParams")
    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        LinkedHashMap<String, ListItemEnum> listMap = createListItemMap();
        List<String> _itemList;
        _itemList = new ArrayList<>(listMap.keySet());

        if (MainActivity.Companion.getListFilterStatus()) {
            _itemList = NyaUtilKt.exclude(_itemList, ListItemEnum.SAFE_MODE.getLocalizedDisplayName(context),
                    ListItemEnum.SOFT_REBOOT.getLocalizedDisplayName(context));
        }

        final String[] items = _itemList.toArray(new String[0]);

        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(context);
        builder.setTitle(R.string.app_name);
        builder.setSingleChoiceItems(items, checkedItem, (dialog, which) -> checkedItem = which);
        builder.setPositiveButton(R.string.confirm, null);
        builder.setNegativeButton(R.string.close, null);
        builder.setNeutralButton(R.string.setting, null);

        AlertDialog dialog = builder.create();
        dialog.setOnShowListener(dialogInterface -> {
            Button positiveButton = dialog.getButton(AlertDialog.BUTTON_POSITIVE);
            Button neutralButton = dialog.getButton(AlertDialog.BUTTON_NEUTRAL);
            // 不调用dismiss()对话框就不会关闭
            positiveButton.setOnClickListener(v ->
                    doAction(ListItemEnum.Companion.fromLocalizedDisplayName(context, items[checkedItem])));
            neutralButton.setOnClickListener(v -> {
//                NavController navController = Navigation.findNavController(requireActivity(), R.id.nav_host_fragment_content_main);
//                navController.navigate(R.id.nav_settings);
                Intent intent = new Intent(context, SettingsActivity.class);
                context.startActivity(intent);
            });
        });

        MaterialAlertDialogBuilder loadingDialog = new MaterialAlertDialogBuilder(context);
        loadingDialog.setTitle(R.string.app_name)
                .setView(getLayoutInflater()
                        .inflate(R.layout.dialog_progress, null))
                .setCancelable(false);

        Intent intent = requireActivity().getIntent();
        if (intent != null && Intent.ACTION_RUN.equals(intent.getAction())) {
            String data = intent.getStringExtra("extra");
            if (data != null)
                new Handler(Looper.getMainLooper()).postDelayed(() -> {
                    doAction(ListItemEnum.Companion.fromDisplayName(data));
                    dismiss();
                }, 1000);
            return loadingDialog.create();
        } else return dialog;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        context = requireActivity();
        rootUtil = new RootUtil(context);
        shizukuUtil = new ShizukuUtil(context);
    }

    @Override
    public void onResume() {
        super.onResume();
        if (NyaSettings.getShizukuShellMode() == NyaSettings.STORE.USER_SERVICE)
            NyaShellManager.INSTANCE.bindService(shizukuUtil, (exitCode, message) -> Log.d("main", "bind " + exitCode));
    }

    private LinkedHashMap<String, ListItemEnum> createListItemMap() {
        LinkedHashMap<String, ListItemEnum> map = new LinkedHashMap<>();
        for (ListItemEnum item : ListItemEnum.getEntries()) {
            map.put(item.getLocalizedDisplayName(context), item);
        }
        return map;
    }

    private void doAction(ListItemEnum listItemEnum) {
        if (NyaSettings.getWorkMode() == NyaSettings.STORE.ROOT)
            funcRoot(listItemEnum);
        else {
            try {
                funcShizuku(listItemEnum);
            } catch (IOException e) {
                e.fillInStackTrace();
                Toast.makeText(context, R.string.exec_fail, Toast.LENGTH_SHORT).show();
            }
        }
    }

    private void runRootCommand(String cmd) {
        if (rootUtil.runRootCommandWithResult(cmd))
            dismiss();
        else
            Toast.makeText(context, R.string.exec_fail, Toast.LENGTH_SHORT).show();
    }

    private void funcRoot(ListItemEnum listItemEnum) {
        switch (listItemEnum) {
            case LOCK_SCREEN:
                runRootCommand("input keyevent KEYCODE_POWER");
                break;
            case REBOOT:
                runRootCommand("svc power reboot");
                break;
            case SOFT_REBOOT:
                runRootCommand("setprop ctl.restart zygote");
                break;
            case SYSTEM_UI:
                runRootCommand("pkill -f com.android.systemui");
                break;
            case RECOVERY:
                runRootCommand("svc power reboot recovery");
                break;
            case BOOTLOADER:
                runRootCommand("svc power reboot bootloader");
                break;
            case SAFE_MODE:
                if (rootUtil.runRootCommandWithResult("setprop persist.sys.safemode 1"))
                    runRootCommand("svc power reboot");
                break;
            case POWER_OFF:
                runRootCommand("reboot -p");
                break;
        }
    }

    private void funcShizuku(ListItemEnum listItemEnum) throws IOException {
        if (!shizukuUtil.checkShizukuPermission()) {
            Toast.makeText(context, R.string.shizuku_denied, Toast.LENGTH_SHORT).show();
            return;
        }
        switch (listItemEnum) {
            case LOCK_SCREEN:
                int lockExitCode = shizukuUtil.runShizukuCommand(new String[]{"input", "keyevent", "KEYCODE_POWER"}, false);
                if (lockExitCode == 0) dismiss();
                break;
            case REBOOT:
                shizukuUtil.shizukuReboot(null);
                break;
            case SOFT_REBOOT:
                shizukuUtil.runShizukuCommand(new String[]{"setprop", "ctl.restart", "zygote"}, true);
                break;
            case SYSTEM_UI:
                int stopUiCode = shizukuUtil.runShizukuCommand(new String[]{"am", "force-stop", "com.android.systemui"}, false);
                if (stopUiCode == 0) {
                    Toast.makeText(context, R.string.stop_system_ui_tip, Toast.LENGTH_LONG).show();
                    dismiss();
                }
                break;
            case RECOVERY:
                shizukuUtil.runShizukuCommand(new String[]{"reboot", "recovery"}, false);
                break;
            case BOOTLOADER:
                shizukuUtil.shizukuReboot("bootloader");
                break;
            case SAFE_MODE:
                int exitCode = shizukuUtil.runShizukuCommand(new String[]{"setprop", "persist.sys.safemode", "1"}, true);
                if (exitCode == 0) shizukuUtil.shizukuReboot(null);
                else Toast.makeText(context, R.string.exec_fail, Toast.LENGTH_SHORT).show();
                break;
            case POWER_OFF:
                shizukuUtil.runShizukuCommand(new String[]{"reboot", "-p"}, false);
                break;
        }
    }

    @Override
    public void onDismiss(@NonNull DialogInterface dialog) {
        super.onDismiss(dialog);
        /*
        为什么使用 System.exit(0) ?
        因为 KernelSU 授权后需要杀死并重启进程使权限生效。
        通常 activity.finish() 仅限于关闭活动，进程由系统决定回收。
        对于像本项目这样的单线程应用，这种做法是安全的。
         */
        NyaShellManager.INSTANCE.unbindService();
        if (!requireActivity().isChangingConfigurations())
            System.exit(0);
    }

}

File: app/src/main/java/github/daisukikaffuchino/rebootnya/fragment/HomeFragment.kt
package github.daisukikaffuchino.rebootnya.fragment

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.ButtonBarLayout
import androidx.fragment.app.DialogFragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import github.daisukikaffuchino.rebootnya.MainActivity
import github.daisukikaffuchino.rebootnya.MainActivity.Companion.EXTRA_ACTION_ITEM
import github.daisukikaffuchino.rebootnya.R
import github.daisukikaffuchino.rebootnya.SettingsActivity
import github.daisukikaffuchino.rebootnya.adapter.HomeRecyclerAdapter
import github.daisukikaffuchino.rebootnya.data.HomeListItemData
import github.daisukikaffuchino.rebootnya.data.ListItemEnum
import github.daisukikaffuchino.rebootnya.shizuku.NyaShellManager
import github.daisukikaffuchino.rebootnya.utils.NyaSettings
import github.daisukikaffuchino.rebootnya.utils.RootUtil
import github.daisukikaffuchino.rebootnya.utils.ShizukuUtil
import github.daisukikaffuchino.rebootnya.utils.isSamsung
import kotlin.system.exitProcess

class HomeFragment : DialogFragment() {
    companion object {
        private val HIDDEN_FOR_LIMITED_SHIZUKU = setOf(
            ListItemEnum.SAFE_MODE,
            ListItemEnum.SOFT_REBOOT
        )
    }

    private lateinit var mContext: Context
    private var checkedItem = 0
    private lateinit var rootUtil: RootUtil
    private lateinit var shizukuUtil: ShizukuUtil

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val intent = requireActivity().intent
        if (intent != null && Intent.ACTION_RUN == intent.action) {
            return createLoadingDialog(intent)
        }

        return if (NyaSettings.getMainInterfaceStyle() == NyaSettings.STYLE.MATERIAL_BUTTON) {
            createMaterialButtonsDialog()
        } else {
            createClassicListDialog()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mContext = requireActivity()
        rootUtil = RootUtil(mContext)
        shizukuUtil = ShizukuUtil(mContext)
    }

    override fun onResume() {
        super.onResume()
        if (NyaSettings.getShizukuShellMode() == NyaSettings.MODE.USER_SERVICE) {
            NyaShellManager.bindService(shizukuUtil) { exitCode, _ ->
                Log.d("main", "bind $exitCode")
            }
        }
    }

    private fun createClassicListDialog(): Dialog {
        val items = prepareDisplayItems()
        val labels = items.map { it.getLocalizedDisplayName(mContext) }.toTypedArray()

        val dialog = MaterialAlertDialogBuilder(mContext)
            .setTitle(R.string.app_name)
            .setSingleChoiceItems(labels, checkedItem) { _, which -> checkedItem = which }
            .create()

        return setupDialogButtons(dialog, items)
    }

    @SuppressLint("InflateParams", "StringFormatInvalid")
    private fun createMaterialButtonsDialog(): Dialog {
        val items = prepareDisplayItems()
        val recyclerView = layoutInflater.inflate(
            R.layout.fragment_home_recycler_view,
            null,
            false
        ) as RecyclerView
        recyclerView.layoutManager = LinearLayoutManager(mContext)

        val data = items.mapIndexed { index, item ->
            HomeListItemData(
                text = item.getLocalizedDisplayName(mContext),
                indexInSection = index,
                sectionCount = items.size,
                checked = index == checkedItem
            )
        }.toMutableList()

        recyclerView.adapter = HomeRecyclerAdapter(data) { _, item ->
            checkedItem = item.indexInSection
        }
        recyclerView.addItemDecoration(HomeRecyclerAdapter.MarginItemDecoration())

        val dialog = MaterialAlertDialogBuilder(mContext)
            .setCustomTitle(layoutInflater.inflate(R.layout.dialog_custom_title, null))
            .setView(recyclerView)
            .create()

        return setupDialogButtons(dialog, items)
    }

    @SuppressLint("InflateParams")
    private fun createLoadingDialog(intent: Intent): Dialog {
        val loadingDialog = MaterialAlertDialogBuilder(mContext)
            .setTitle(R.string.app_name)
            .setView(layoutInflater.inflate(R.layout.dialog_progress, null))
            .setCancelable(false)

        intent.getStringExtra(EXTRA_ACTION_ITEM)?.let { displayName ->
            Handler(Looper.getMainLooper()).postDelayed({
                doAction(ListItemEnum.fromDisplayName(displayName))
                dismiss()
            }, 1000)
        }

        return loadingDialog.create()
    }

    @SuppressLint("RestrictedApi")
    private fun setupDialogButtons(
        dialog: AlertDialog,
        items: List<ListItemEnum>
    ): AlertDialog {
        dialog.setButton(AlertDialog.BUTTON_POSITIVE, getString(R.string.confirm)) { _, _ -> }
        dialog.setButton(AlertDialog.BUTTON_NEGATIVE, getString(R.string.close)) { _, _ -> }
        dialog.setButton(AlertDialog.BUTTON_NEUTRAL, getString(R.string.setting)) { _, _ -> }

        dialog.setOnShowListener {
            val positiveButton = dialog.getButton(AlertDialog.BUTTON_POSITIVE)
            val neutralButton = dialog.getButton(AlertDialog.BUTTON_NEUTRAL)

            if (NyaSettings.getMainInterfaceStyle() == NyaSettings.STYLE.CLASSIC_LIST) {
                dialog.listView.setItemChecked(checkedItem, true)
                dialog.listView.setSelection(checkedItem)
            }

            positiveButton.setOnClickListener {
                val itemEnum = items.getOrNull(checkedItem) ?: return@setOnClickListener
                NyaSettings.setLastSelectedOption(itemEnum.name)
                doAction(itemEnum)
            }
            neutralButton.setOnClickListener {
                mContext.startActivity(Intent(mContext, SettingsActivity::class.java))
            }

            findButtonBarLayout(dialog.window?.decorView)?.setAllowStacking(false)
        }
        return dialog
    }

    @SuppressLint("RestrictedApi")
    private fun findButtonBarLayout(root: View?): ButtonBarLayout? {
        if (root == null) return null
        if (root is ButtonBarLayout) return root
        if (root is ViewGroup) {
            for (i in 0 until root.childCount) {
                findButtonBarLayout(root.getChildAt(i))?.let { return it }
            }
        }
        return null
    }

    private fun prepareDisplayItems(): List<ListItemEnum> {
        val items = getAvailableItems()
        checkedItem = resolveCheckedItem(items)
        return items
    }

    private fun getAvailableItems(): List<ListItemEnum> {
        val shizukuFilterEnabled = MainActivity.checkListFilterStatus()
        val samsungDevice = isSamsung()

        return ListItemEnum.entries.filterNot { item ->
            (shizukuFilterEnabled && item in HIDDEN_FOR_LIMITED_SHIZUKU) ||
                    (samsungDevice && item == ListItemEnum.BOOTLOADER) ||
                    (!samsungDevice && item == ListItemEnum.SAMSUNG_DOWNLOAD)
        }
    }

    private fun resolveCheckedItem(items: List<ListItemEnum>): Int {
        val lastSelected = NyaSettings.getLastSelectedOption()
        val selectedIndex = items.indexOfFirst { it.name == lastSelected }
        if (selectedIndex >= 0) {
            return selectedIndex
        }

        if (lastSelected != null) {
            items.firstOrNull()?.let { NyaSettings.setLastSelectedOption(it.name) }
        }
        return 0
    }

    private fun doAction(listItemEnum: ListItemEnum) {
        if (NyaSettings.getWorkMode() == NyaSettings.MODE.ROOT) {
            funcRoot(listItemEnum)
        } else {
            funcShizuku(listItemEnum)
        }
    }

    private fun runRootCommand(cmd: String) {
        if (rootUtil.runRootCommandWithResult(cmd)) {
            dismiss()
        } else {
            Toast.makeText(mContext, R.string.exec_fail, Toast.LENGTH_SHORT).show()
        }
    }

    private fun funcRoot(listItemEnum: ListItemEnum) {
        when (listItemEnum) {
            ListItemEnum.LOCK_SCREEN -> runRootCommand("input keyevent KEYCODE_POWER")
            ListItemEnum.REBOOT -> runRootCommand("svc power reboot")
            ListItemEnum.SOFT_REBOOT -> runRootCommand("setprop ctl.restart zygote")
            ListItemEnum.SYSTEM_UI -> runRootCommand("pkill -f com.android.systemui")
            ListItemEnum.RECOVERY -> runRootCommand("svc power reboot recovery")
            ListItemEnum.BOOTLOADER -> runRootCommand("svc power reboot bootloader")
            ListItemEnum.SAMSUNG_DOWNLOAD -> runRootCommand("reboot download")
            ListItemEnum.SAFE_MODE -> {
                if (rootUtil.runRootCommandWithResult("setprop persist.sys.safemode 1")) {
                    runRootCommand("svc power reboot")
                }
            }

            ListItemEnum.POWER_OFF -> runRootCommand("reboot -p")
        }
    }

    private fun funcShizuku(listItemEnum: ListItemEnum) {
        if (!shizukuUtil.checkShizukuPermission()) {
            Toast.makeText(mContext, R.string.shizuku_denied, Toast.LENGTH_SHORT).show()
            return
        }

        when (listItemEnum) {
            ListItemEnum.LOCK_SCREEN -> {
                val lockExitCode = shizukuUtil.runShizukuCommand(
                    arrayOf("input", "keyevent", "KEYCODE_POWER"),
                    false
                )
                if (lockExitCode == 0) {
                    dismiss()
                }
            }

            ListItemEnum.REBOOT -> shizukuUtil.shizukuReboot(null)
            ListItemEnum.SOFT_REBOOT -> shizukuUtil.runShizukuCommand(
                arrayOf("setprop", "ctl.restart", "zygote"),
                true
            )

            ListItemEnum.SYSTEM_UI -> {
                val stopUiCode = shizukuUtil.runShizukuCommand(
                    arrayOf("am", "force-stop", "com.android.systemui"),
                    false
                )
                if (stopUiCode == 0) {
                    Toast.makeText(mContext, R.string.stop_system_ui_tip, Toast.LENGTH_LONG).show()
                    dismiss()
                }
            }

            ListItemEnum.RECOVERY -> shizukuUtil.runShizukuCommand(
                arrayOf("reboot", "recovery"),
                false
            )

            ListItemEnum.BOOTLOADER -> shizukuUtil.shizukuReboot("bootloader")

            ListItemEnum.SAMSUNG_DOWNLOAD -> shizukuUtil.runShizukuCommand(
                arrayOf("reboot", "download"),
                false
            )

            ListItemEnum.SAFE_MODE -> {
                val exitCode = shizukuUtil.runShizukuCommand(
                    arrayOf("setprop", "persist.sys.safemode", "1"),
                    true
                )
                if (exitCode == 0) {
                    shizukuUtil.shizukuReboot(null)
                } else {
                    Toast.makeText(mContext, R.string.exec_fail, Toast.LENGTH_SHORT).show()
                }
            }

            ListItemEnum.POWER_OFF -> {
                shizukuUtil.runShizukuCommand(arrayOf("reboot", "-p"), false)
            }
        }
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        NyaShellManager.unbindService()
        if (!requireActivity().isChangingConfigurations) {
            exitProcess(0)
        }
    }
}
